### ISSUE TITLE
bug: child configs that use 'select' or 'extend-select' invalidate the parent config's 'ignore' list.

### ISSUE THREAD

**Issue by Hnasar** [astral-sh/ruff#10622]
## Bug/Use-case

As someone who helps maintain a monorepo with many subprojects and child configs, I want to be able to maintain a centralized configuration of ignored rules (which projects may override with `tool.ruff.extend`). There appears to be a bug where the top-level `tool.ruff.lint.ignore` configuration _isn't honored_ whenever a child config modifies either `tool.ruff.lint.select` or `tool.ruff.lint.extend-select`. This occurs all the time, even if the child config HAS NOT overridden `tool.ruff.lint.ignore` in any way.

I (think I) understand the semantics of `ruff.extend`  alongside `lint.extend-select` and `lint.select` but the semantics of the now-combined `lint.select` create some really confusing behaviors, and mean I don't have a good way to fulfill my use-case mentioned above.

## Repro

Using ruff 0.3.4

```console
$ tree
.
└── parent
    ├── child
    │   ├── foo.py
    │   └── pyproject.toml
    └── pyproject.toml
```

```py
# child/foo.py
import os, re

"{}".format(1)
```

### 0. Parent ignores with child select

```toml
# parent/pyproject.toml
[tool.ruff]
lint.select = ["E", "UP"]
lint.ignore = ["E401", "UP032"]
```
```toml
# child/pyproject.toml
[tool.ruff]
extend = "../pyproject.toml"
lint.extend-select = ["UP"]  # also same behavior if the we use lint.select ["UP"]
```

**Observed**: 
```console
$ ruff check .
parent/child/foo.py:3:1: UP032 [*] Use f-string instead of `format` call
```

**Expected**:
```console
$ ruff check .
All checks passed!
```

* I expect that if I want the child config to invalidate the parent `ignore` list, then I have to provide my own `ignore` list. 

### 1. Parent ignores with child select and child ignore

```toml
# parent/pyproject.toml
[tool.ruff]
lint.select = ["E", "UP"]
lint.ignore = ["E401", "UP032"]
```
```toml
# child/pyproject.toml
[tool.ruff]
extend = "../pyproject.toml"
lint.extend-select = ["UP"]  # also same behavior if the we use lint.select ["UP"]
lint.ignore = ["UP032"]
```

**Observed**: 
```console
$ ruff check .
All checks passed!
```

**Expected**:
```console
$ ruff check .
parent/child/foo.py:1:1: E401 [*] Multiple imports on one line
```

* And I expect that if I provide a child `ignore` list then it fully invalidates the parent `ignore` list.

## Workarounds

Neither of these are ideal, but are some things I tried:

1. Copy and paste the parent's `lint.ignore` and keep it in sync in all the child configs. This is cumbersome to keep in sync and easy to miss things.
2. Instead of using `select` or `extend-select` in child configs, I can invert this and `ignore` all the top-level `selects` that a project wants to opt-out of. This is confusing, and also is cumbersome and easy to miss.

## Examples of fine behavior that makes sense
<details>
  <summary>Click to expand examples</summary>
  

### 2. No ignores

```toml
# parent/pyproject.toml
[tool.ruff]
lint.select = ["E", "UP"]
```

```toml
# child/pyproject.toml
[tool.ruff]
extend = "../pyproject.toml"
```

```console
$ ruff  check .
parent/child/foo.py:1:1: E401 [*] Multiple imports on one line
parent/child/foo.py:3:1: UP032 [*] Use f-string instead of `format` call
```

### 3. Child select
```toml
# parent/pyproject.toml
[tool.ruff]
lint.select = ["E", "UP"]
```
```toml
# child/pyproject.toml
[tool.ruff]
extend = "../pyproject.toml"
lint.select = ["UP"]
```
```console
$ ruff  check .
parent/child/foo.py:3:1: UP032 [*] Use f-string instead of `format` call
```

### 4. Child extend-select

```toml
# parent/pyproject.toml
[tool.ruff]
lint.select = ["E", "UP"]
```
```toml
# child/pyproject.toml
[tool.ruff]
extend = "../pyproject.toml"
lint.extend-select = ["UP"]
```

```console
$ ruff  check .
parent/child/foo.py:1:1: E401 [*] Multiple imports on one line
parent/child/foo.py:3:1: UP032 [*] Use f-string instead of `format` call
```

### 5. Parent ignores without any child select

```toml
# parent/pyproject.toml
[tool.ruff]
lint.select = ["E", "UP"]
lint.ignore = ["E401", "UP032"]
```
```toml
# child/pyproject.toml
[tool.ruff]
extend = "../pyproject.toml"
```
```console
ruff  check .
All checks passed!
```
</details>

**Comment by charliermarsh** [comment]
So for "0. Parent ignores with child select", to the best of my recollection, I thought a child `select` was intended to "reset" the rule selection (which tends to create more intuitive behavior), but an `extend-select` was _not_ supposed to reset. So your example there _should_ work as you expected. I'll have to test it out myself and see what's up.

**Comment by MichaReiser** [comment]
Here's a ZIP with the described repo. 
[I10622.zip](https://github.com/astral-sh/ruff/files/14770373/I10622.zip)

**Comment by charliermarsh** [comment]
I can take a look at this tonight.

**Comment by charliermarsh** [comment]
On further review this is working as expected but my previous explanation was missing one piece...

When you use `extends`, we resolve the list of resolves one configuration file at a time. So, we first resolve all of the selectors in `parent/pyproject.toml`, and come up with a list of rules. Then, we apply the selectors in `child/pyproject.toml` on top of that list. So if you have `lint.extend-select = ["UP"]` in your `child/pyproject.toml`, it will indeed then add all the `UP` rules and ignore the `UP032` ignore from `parent/pyproject.toml`. However, note that `E401` is _still_ disabled.

**Comment by Hnasar** [comment]
Thanks for the explanation of how it works with the file-by-file resolution, because I was really confused because I thought the 'most specific option wins' was the standard semantics.

So then for my initial user story, is there any supported way to globally ignore some rules when there's a tree of configs? If we had a `tool.ruff.lint.ignore` that overrode any inherited ignores, and a `tool.ruff.lint.extend-ignore` which _added_ to any inherited ignores, that would give me the tools I need to fix things, but otherwise I'm not sure.

**Comment by MichaReiser** [comment]
I think it would be good for us to document this behavior because it wasn't clear to me.

**Comment by Avasam** [comment]
This would be quite problematic with shared configs: https://github.com/astral-sh/ruff/issues/12352

Edit: now ignore and extend-ignore are the same.
I would expect "select" to "reset" the parent's selection. And "extend-select" to expand upon it. Which I believe is currently the case.

**Comment by cscanlin** [comment]
I want to bump this one up because this did not work at all how I expect, and I'm wondering if what I want to do is even possible. I have a more detailed set of rules that I want to use (for e.g. my text editor) on all projects that is more restrictive than what each project itself enforces.

What I originally tried (simplified):
```
# project/pyproject.toml
[tool.ruff]
extend = "~/pyproject.toml"
lint.select = ["E101"]
```

```
# ~/pyproject.toml
[tool.ruff]
lint.extend-select = ["F701"]
```

When I run `ruff check . --show-settings`, I get the following rules:
```
linter.rules.enabled = [
	mixed-spaces-and-tabs (E101),
]
```
I would also expect F701 to show up there.

Overriding the extend-select does work as expected:
`ruff check . --show-settings --extend-select E401`
```
linter.rules.enabled = [
	mixed-spaces-and-tabs (E101),
	multiple-imports-on-one-line (E401),
]
```

Interestingly, if I flip this around so the select is in the parent it kind of works
```
# project/pyproject.toml
[tool.ruff]
extend = "~/pyproject.toml"
lint.extend-select = ["E101"]
```

```
# ~/pyproject.toml
[tool.ruff]
lint.select = ["F701"]
```

```
linter.rules.enabled = [
	mixed-spaces-and-tabs (E101),
	break-outside-loop (F701),
]
```

But this will affect all instances of when ruff is run in my environment, so I'd like to be able to override the parent config select in certain cases, e.g.:
`ruff check . --show-settings --select E401`

And I get the following:
```
linter.rules.enabled = [
	multiple-imports-on-one-line (E401),
]
```

I would have expected to E101 to be included here, since the extend-select is never overridden.

**Comment by Avasam** [comment]
> I would also expect F701 to show up there.

@cscanlin I wouldn't, your examples seem to be working as I would expect.

`select` will make it so *only* those listed rules are selected. You want to use `extend-select` in all cases here.
